core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,403 @@
+                    state_stats_mapping[state_name].first_hit_count_v1 += 1
+                    state_session_ids[state_name].append(value['session_id'])
+                    session_ids_set.add(value['session_id'])
+
+            # Compute num_completions for the states.
+            for session_id in session_ids_set:
+                values_state_hit_sessioned = [
+                    val for val in values_state_hit_versioned if val[
+                        'session_id'] == session_id]
+                # All states in the above list except the last element (only if
+                # complete exploration event doesn't exist for the session ID)
+                # are counted as completed because they are ordered by their
+                # timestamp.
+                if session_id not in completed_session_ids:
+                    for value in values_state_hit_sessioned[:-1]:
+                        state_stats_mapping[value[
+                            'state_name']].num_completions_v1 += 1
+                else:
+                    for value in values_state_hit_sessioned:
+                        state_stats_mapping[value[
+                            'state_name']].num_completions_v1 += 1
+
+            # Compute total_answers_count and useful_feedback_count.
+            for value in values_state_answer_versioned:
+                for answer in value['submitted_answer_list']:
+                    state_stats_mapping[value[
+                        'state_name']].total_answers_count_v1 += 1
+                    if answer['classification_categorization'] != (
+                            exp_domain.DEFAULT_OUTCOME_CLASSIFICATION):
+                        state_stats_mapping[value[
+                            'state_name']].useful_feedback_count_v1 += 1
+
+        # Now that the stats for the complete exploration are computed,
+        # calculate num_actual_starts.
+        num_actual_starts = state_stats_mapping[
+            exploration.states[
+                exploration.init_state_name].interaction.answer_groups[
+                    0].outcome.dest].first_hit_count_v1
+
+        # Create the ExplorationStatsModel instance.
+        exploration_stats = stats_domain.ExplorationStats(
+            exp_id, latest_exp_version, num_starts, 0, num_actual_starts, 0,
+            num_completions, 0, state_stats_mapping)
+        stats_services.create_stats_model(exploration_stats)
+
+
+class StatisticsAuditV2(jobs.BaseMapReduceOneOffJobManager):
+    """A one-off statistics audit.
+
+    Performs a brief audit of exploration stats values to ensure that they
+    maintain simple sanity checks like being non-negative and so on.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [stats_models.ExplorationStatsModel]
+
+    @classmethod
+    def require_non_negative(
+            cls, exp_id, property_name, value, state_name=None):
+        if not state_name:
```

Comment:
Too much duplication:

(a) Pass in just the numerical value of the stat, and the version, separately. So: cls, exp_id, exp_version, property_name, value, state_name=None
(b) Find a way to scope the if/else down, or remove it altogether. Maybe use the second string but have the fourth argument be `... if state_name else ''`.